### ISSUE TITLE
On testing the CLI

### ISSUE THREAD

**Issue by ErisDS** [TryGhost/Ghost-CLI#328]
This isn't something I'm really expecting to be looked at asap, however I've had a couple of stabs at getting more tests in and have really struggled. I wanted to write down what I've learned whilst it's fresh so that someone else, or me, can pick up where I left off.

Scroll down for a TLDR!!!!!

----

**What I wanted to achieve**

I was trying to add a single acceptance test for `ghost install local`. Ostensibly, this one command should work anywhere, meaning we can write one test and just run it on travis, locally + appveyor and know that it works on linux, mac and windows.

This is cheating, but it is a start. I thought it would be a great additional acceptance test. I wouldn't want too many, this one would take ~35seconds.

My attempt can be found here: https://github.com/ErisDS/Ghost-CLI/commit/a28f3944c22a86f238232caf26909b35b17fe44c

I ran into a list of challenges:

- There is an outline of a framework for acceptance testing in place, but only minimal usage / docs to be read
- It currently only supports `test.setup('full')` which fakes an install and is useful for the config acceptance tests
- I tried to add `test.setup('empty')` which should be an empty directory managed by tmp
- I didn't know whether to use `test.run` or `test.spawn` - can't figure out which one is best for what
- The promises returned by `test.x` are not bluebird, so finally didn't work
- Fixed this with the global.Promise = require('bluebird') hack, which we should probably use anyway
- There is a function to cleanup the tmp directory, but it is never called, meaning even the existing tests will be filling up your harddrive!
- The test I wrote works if I comment out the .finally() block, but this results in Ghost being left installed and running 😱
- With the .finally() block in place, I get this output:

```
Error: Command failed: /Users/hannah/Ghost/Ghost-CLI/bin/ghost install local
Message: Current directory is not empty, Ghost cannot be installed here.

'ghost install local' failed
cwd: undefined
stdout:
 A SystemError occured.

Debug Information:
    Node Version: v6.10.3
    Ghost-CLI Version: 1.0.0-beta.6
    Environment: production
    Command: 'ghost install local'

Please refer to https://docs.ghost.org/docs/installing-ghost-via-the-cli#troubleshooting for troubleshooting.
 
stderr: Message: Current directory is not empty, Ghost cannot be installed here.


      at ChildProcess.exithandler (child_process.js:204:12)
      at maybeClose (internal/child_process.js:886:16)
      at Socket.<anonymous> (internal/child_process.js:342:11)
      at Pipe._handle.close [as _onclose] (net.js:497:12)
```

----

**Ghost CLI testing challenges**

Other CLIs have similar challenges for testing, and the framework in place, borrowed from denali, attempts to cover some of them.

E.g. the `env` tooling handles setting up a file, directory & symlink environment
E.g. `test.spawn` attempts to handle stopping the command that is run

However, Ghost CLI has additional challenges:

- It depends on the system environment, far beyond the local files, folders and symlinks. E.g. OS, permissions, installed software, etc will all impact on the CLI's behaviour
- Ghost CLI spawns its own processes (Ghost), which also have to be managed (stopped on test pass or test fail).
- Ghost CLI attempts to create files and directories outside of the CWD. 

These 3 things would also need to be managed by any acceptance tests. 

The amount of work involved here is pretty huge.

-----

**What DOES need testing?** 

This brings me full circle back to the question of whether the test I was trying to write is even useful, and by reflecting on the bugs and regressions we've experienced over the last couple of weeks, I think there are a couple of clear things that ought to be focused upon for testing.

I believe that focusing on these things could be done with unit testing alone, removing the need for acceptance tests. I think the config tests could also be rewritten as unit tests, given that `tmp` is used to manage the environment in which it is run.

- `argv` config for each command:
   - when a command is run without flags, does "argv" get the correct defaults?
   - when a command is run with flags, does "argv" get the correct values?
- file & folder permissions: 
   - whenever ghost-cli attempts to create, read, or update a file, are the permissions correct before and after?
  - do the commands that rely on folder permissions (e.g. `ghost log` work when the permissions are set to what we _expect_ they should be (e.g. again using [tmp](https://github.com/raszi/node-tmp) to control the folder

A slightly different approach is looking at open issues and thinking about how we might test them, E.g.  #304 Skip MySQL extension if sqlite was configured

This forces us to think a little bit about the layered approach of `ghost install`...

- What does `ghost doctor` do when passed `--db=sqlite3`
- What does `ghost config` do when passed `--db=sqlite3`
- What does `ghost setup` do when passed `--db=sqlite3`
- What does `ghost install` do when passed `--db=sqlite3`

I think each command can be tested using the pattern in the [restart](https://github.com/TryGhost/Ghost-CLI/blob/3e17cfe975565467d9909affc28b26a27eb00cf3/test/unit/commands/restart.js#L21-L21) tests?

From this approach, I believe the main testing tools we need to perform deeper command checks are:
1.  an easy way to create a test "instance" as in this.system.getInstance(), as shown in the Restart command test - this could be abstracted out to be part of our unit testing framework
2. a tool for faking execa, so that execa returns stdout/stderr results similar to what it would IRL
3. a test version of this.ui, which is probably just a set of spies?!

On top of this, in the code itself, we will probably need to group together some similar code into functions, to make it easier to stub/fake.

E.g. The checks for ghost doctor area almost all direct calls to execa, which is fine, but it would probably be easier to test if there was a checks.js or env.js with functions like `checkMySQLExists()`.

E.g. All code that uses fs-extra being grouped into a `file-manager.js` for easier stubbing and testing.

----

# TLDR;

I [tried to implement](https://github.com/ErisDS/Ghost-CLI/commit/a28f3944c22a86f238232caf26909b35b17fe44c) an acceptance test for `ghost install local` and it didn't work for many complex reasons.

I think we can test all of the behaviour of Ghost-CLI with unit testing only. 
I think we need a couple of small tools and refactors to make this easy.

Small TODOs:

- [ ] existing acceptance tests need to call cleanupDir (or be deleted)
- [ ] consider using the global bluebird promise override
